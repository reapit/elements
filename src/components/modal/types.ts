import { ForwardRefExoticComponent, HTMLAttributes, RefAttributes } from 'react'

export interface ModalProps extends HTMLAttributes<HTMLDivElement> {
  isOpen: boolean
  onModalClose: () => void
  title?: string
  className?: string
}

export interface ModalBaseProps extends HTMLAttributes<HTMLElement> {}

export type ModalContainerProps = ForwardRefExoticComponent<ModalBaseProps & RefAttributes<HTMLAttributes<HTMLElement>>>
