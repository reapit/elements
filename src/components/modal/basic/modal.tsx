import { FC } from 'react'
import { cx } from '@linaria/core'
import { ElModalBg, ElModal, ElModalHeader } from '../__styles__'
import { ModalBaseProps } from '../types'

export const ModalBg: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalBg className={cx(className)} {...rest}>
    {children}
  </ElModalBg>
)

export const ModalContainer: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModal className={cx(className)} {...rest}>
    {children}
  </ElModal>
)

export const ModalHeader: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalHeader className={cx(className)} {...rest}>
    {children}
  </ElModalHeader>
)

export const ModalBody: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalBg className={cx(className)} {...rest}>
    {children}
  </ElModalBg>
)
