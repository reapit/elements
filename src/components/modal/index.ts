export * from './styles'
export * from './types'
export * from './modal.atoms'
export * from './modal'
