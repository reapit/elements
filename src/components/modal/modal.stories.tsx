import { useState } from 'react'
import { useModal } from '../../hooks/use-modal'
import { Modal, ModalBg } from './index'
import { Button, ButtonGroup } from '../button'
import { BodyText } from '../typography'

export default {
  title: 'Modal',
  component: Modal,
}

export const BasicUsageClosed = {
  render: ({}) => <ModalBg />,
  name: 'Basic Usage - Closed',
}

export const ReactModal = {
  render: ({}) => {
    const [modalIsOpen, setModalIsOpen] = useState(false)

    return (
      <>
        <Button intent="primary" onClick={() => setModalIsOpen(!modalIsOpen)}>
          Open Modal
        </Button>
        <Modal isOpen={modalIsOpen} onModalClose={() => setModalIsOpen(!modalIsOpen)} title="Welcome to the demo modal">
          <BodyText>Here&apos;s some nice content for the inside of the modal.</BodyText>
        </Modal>
      </>
    )
  },
}

export const UseModalExample = {
  render: ({}) => {
    const { Modal: ModalComponent, openModal, closeModal } = useModal('storybook-docs')

    return (
      <>
        <Button intent="primary" onClick={openModal}>
          Open Modal
        </Button>
        <ModalComponent title="Welcome to the demo modal">
          <BodyText>Here&apos;s some nice content for the inside of the modal.</BodyText>
          <ButtonGroup alignment="right">
            <Button intent="default" onClick={closeModal}>
              Close
            </Button>
            <Button intent="primary" onClick={console.log}>
              Do Something
            </Button>
          </ButtonGroup>
        </ModalComponent>
      </>
    )
  },

  name: 'useModal example',
}

export const UseModalExampleMultipleModals = {
  render: ({}) => {
    const { Modal: ModalA, openModal: openModalA } = useModal('storybook-docs')

    const { Modal: ModalB, openModal: openModalB } = useModal('storybook-docs')

    return (
      <>
        <ButtonGroup>
          <Button intent="primary" onClick={openModalA}>
            Open Modal A
          </Button>
          <Button intent="primary" onClick={openModalB}>
            Open Modal B
          </Button>
        </ButtonGroup>
        <ModalA title="Modal A">I&apos;m the modal A&apos;s content</ModalA>
        <ModalB title="Modal B">I&apos;m the modal B&apos;s content</ModalB>
      </>
    )
  },

  name: 'useModal example - multiple modals',
}
