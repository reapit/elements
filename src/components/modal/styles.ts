import { styled } from '@linaria/react'
import { elIsActive } from '../../styles/states'
import { css } from '@linaria/core'

// NOTE: The `elModalSmallVariant` class is used to indicate that the modal
// is a small variant. This class is used to apply the appropriate styles to
// the modal when it is a small variant, and is applied to the `className`
// prop of the `ElModal` component
export const elModalSmallVariant = css``

export const ElModalBg = styled.div`
  display: none;
  z-index: 98;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  background: var(--neutral-500);
  opacity: 0.2;

  &.${elIsActive} {
    display: block;
    position: fixed;
  }
`

export const ElModal = styled.div`
  display: none;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background: white;
  z-index: 99;
  overflow: auto;
  box-shadow: var(--elevation-xl, 0px 4px 16px 0px rgba(96, 120, 144, 0.16));
  border-radius: var(--corner-lg, 8px);
  padding: var(--spacing-5, 20px) var(--spacing-6, 24px);

  &.${elIsActive} {
    display: block;
    position: fixed;
  }

  /* default modal size settings */
  min-width: 300px;
  max-height: 100%;
  width: calc(100% - var(--unit-8, 32px));
  max-width: 30rem;

  /* small modal size settings */
  &.${elModalSmallVariant} {
    width: calc(100% - var(--unit-16, 64px));
    max-width: 20rem;
  }
`

export const ElModalHeader = styled.div`
  color: var(--text-primary, #222b33);
  font-family: var(--font-family, Inter);
  font-size: var(--font-size-lg, 18px);
  font-style: normal;
  font-weight: 600;
  line-height: var(--line-height-lg, 24px);
  letter-spacing: var(--letter-spacing-lg, -0.18px);
  padding-bottom: var(--spacing-4, 12px);
  word-wrap: break-word;
`

export const ElModalBody = styled.div`
  font-size: var(--font-size-default);
`

export const ElModalFooter = styled.div`
  display: flex;
  justify-content: flex-end;
  padding-top: var(--spacing-5, 20px);
`
