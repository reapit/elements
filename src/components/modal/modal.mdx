import { Meta, Story, Canvas, Controls } from '@storybook/blocks'
import { Modal } from './index'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as ModalStories from './modal.stories'

<Meta of={ModalStories} />

# Modal

The modal component renders a dialog box that overlays the main screen. It is typically used for displaying confirmation or error notifications to users that require an addtional call to action or step.

## Basic Usage

The Modal component can be used for UI purposes only, but won't have the functionality of the React implementation (i.e. managing the state of whether it is open or closed)

If using the Modal component like this, you will need to toggle the `elIsActive` / `el-is-active` class to show/hide the modal.

This class must be applied on two elements, `div.el-modal-bg` and also `div.el-modal`, when you want the modal to be shown.

By default, the modal fills it's current container so you will likely wish to place it as high as possible up the DOM tree to achive an overlay effect.

Click "Show code" below to see JSX / HTML code for a closed modal, along with the corresponding rendered markup. To see how an opened one behaves and the markup, move to the React examples below.

<Canvas of={ModalStories.BasicUsageClosed} />

<RenderHtmlMarkup of={ModalStories.BasicUsageClosed} />

## React usage

The React component accepts props to toggle the classes that control whether the modal is opened or not.

In React world, you will likely want to trigger the modal using a portal (see our `Portal` component), so it can be rendered at the top level of your app and provide an overlay effect. Probably you will `useState` to toggle the hide and show behaviour.

This example gives maxmum flexibility but for the 'batteries included' approach, you should look at the `useModal` hook below.

Props available for the React version of the component;

<Controls />

<Canvas of={ModalStories.ReactModal} story={{height: '300px'}} />

## React usage - Complete example with `useModal` hook

The use modal hook exports both a modal component and state helpers for toggling the component open.

It also leverages our `Portal` component to render the modal at the top level of your app. The default container is `#root` so if you need your portal to render on another element, pass a different id to the `useModal` hook when you instantiate it - in the example below, we use Storybook's `#docs-root`.

<Canvas of={ModalStories.UseModalExample} />

## React usage - With multiple `useModal` hooks

Props available for the React version of the component;

If using multiple modals within a component, ensure you rename the exports from `useModal` to avoid conflicts, e.g. `const { Modal: ModalA, openModal: openModalA } = useModal()`

<Canvas of={ModalStories.UseModalExampleMultipleModals} />

## `useModal` API

Exports the following properties as an object:

- `Modal`: `FC<Partial<ModalProps>>` - a React component that should be placed somewhere with the component that you want to use the modal. Will pass props to the underlying modal.
- `openModal`: `() => void` - a method that will open the Modal.
- `closeModal`: `() => void` - a method that will close the Modal. The Modal can also be closed by default from within the Modal, or by clicking on the background
- `modalIsOpen`: `boolean` - convenience boolean to check if the modal is open currently.
