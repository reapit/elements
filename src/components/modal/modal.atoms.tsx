import { FC, ForwardedRef, forwardRef, HTMLAttributes, LegacyRef } from 'react'
import { cx } from '@linaria/core'
import { ElModalBg, ElModal, ElModalHeader } from './styles'
import { ModalBaseProps, ModalContainerProps } from './types'

export const ModalBg: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalBg className={cx(className)} {...rest}>
    {children}
  </ElModalBg>
)

export const ModalContainer: ModalContainerProps = forwardRef(
  ({ className, children, ...rest }, ref: ForwardedRef<HTMLAttributes<HTMLElement>>) => (
    <ElModal className={cx(className)} {...rest} ref={ref as unknown as LegacyRef<HTMLInputElement>}>
      {children}
    </ElModal>
  ),
)

export const ModalHeader: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalHeader className={cx(className)} {...rest}>
    {children}
  </ElModalHeader>
)

export const ModalBody: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalBg className={cx(className)} {...rest}>
    {children}
  </ElModalBg>
)
