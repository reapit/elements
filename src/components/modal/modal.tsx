import React, { FC, HTMLAttributes, createRef, useEffect, useId } from 'react'
import { cx } from '@linaria/core'
import { ElModalBg, ElModal, ElModalHeader, ElModalBody } from './__styles__'
import { elIsActive } from '../../styles/states'

export interface ModalProps extends HTMLAttributes<HTMLDivElement> {
  isOpen: boolean
  onModalClose: () => void
  title?: string
  className?: string
}

export interface ModalBaseProps extends HTMLAttributes<HTMLElement> {}

export const ModalBg: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalBg className={cx(className)} {...rest}>
    {children}
  </ElModalBg>
)

export const ModalContainer: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModal className={cx(className)} {...rest}>
    {children}
  </ElModal>
)

export const ModalHeader: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalHeader className={cx(className)} {...rest}>
    {children}
  </ElModalHeader>
)

export const ModalBody: FC<ModalBaseProps> = ({ className, children, ...rest }: ModalBaseProps) => (
  <ElModalBg className={cx(className)} {...rest}>
    {children}
  </ElModalBg>
)

export const handleModalFocus = (modalRef: React.RefObject<HTMLDivElement>, isOpen: boolean) => () => {
  if (isOpen && modalRef.current) {
    modalRef.current.focus()
  }
}

export const Modal: FC<ModalProps> = ({ isOpen, onModalClose, title, className, children, ...rest }) => {
  const id = rest.id || useId()
  const modalRef = createRef<HTMLDivElement>()

  useEffect(() => {
    const onKeyUp = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        onModalClose()
      }
    }
    document.addEventListener('keyup', onKeyUp, false)

    return () => {
      document.removeEventListener('keyup', onKeyUp, false)
    }
  }, [onModalClose])

  const modalCombinedClassname = cx(className, elIsActive)

  useEffect(handleModalFocus(modalRef, isOpen), [modalRef, isOpen])

  if (!isOpen) return null

  return (
    <>
      <ElModalBg className={elIsActive} onClick={onModalClose} />
      <ElModal
        role="dialog"
        aria-modal="true"
        aria-describedby={id}
        className={modalCombinedClassname}
        ref={modalRef}
        autoFocus
        {...rest}
      >
        {title && <ElModalHeader>{title}</ElModalHeader>}
        <ElModalBody id={id}>{children}</ElModalBody>
      </ElModal>
    </>
  )
}
