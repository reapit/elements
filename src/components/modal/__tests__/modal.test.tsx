import { render } from '@testing-library/react'
import { Modal, handleModalFocus } from '../modal'
import { createRef } from 'react'

describe('Modal component', () => {
  it('should match a snapshot when closed', () => {
    const wrapper = render(
      <Modal isOpen={false} onModalClose={() => {}} title="test">
        Content within modal
      </Modal>,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should match a snapshot when open', () => {
    const wrapper = render(
      <Modal isOpen={true} onModalClose={() => {}} title="test">
        Content within modal
      </Modal>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('handleModalFocus', () => {
  it('should focus on the modal if it is open', () => {
    const modalRef = createRef<HTMLDivElement>()
    const mockElement = document.createElement('div')
    const focusSpy = vi.spyOn(mockElement, 'focus')

    Object.defineProperty(modalRef, 'current', {
      value: mockElement,
    })

    const curried = handleModalFocus(modalRef, true)

    curried()

    expect(focusSpy).toHaveBeenCalled()
  })

  it('should not focus on the modal if it is not open', () => {
    const modalRef = createRef<HTMLDivElement>()
    const mockElement = document.createElement('div')
    const focusSpy = vi.spyOn(mockElement, 'focus')

    Object.defineProperty(modalRef, 'current', {
      value: mockElement,
    })

    const curried = handleModalFocus(modalRef, false)

    curried()

    expect(focusSpy).not.toHaveBeenCalled()
  })
})
