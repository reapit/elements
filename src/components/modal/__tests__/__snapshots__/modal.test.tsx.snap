// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal component > should match a snapshot when closed 1`] = `<DocumentFragment />`;

exports[`Modal component > should match a snapshot when open 1`] = `
<DocumentFragment>
  <div
    class="el-is-active el-modal-bg"
  />
  <div
    aria-describedby="test-static-id"
    aria-modal="true"
    class="el-is-active el-modal"
    role="dialog"
  >
    <div
      class="mocked-styled-2 el-modal-header"
    >
      test
    </div>
    <div
      class="mocked-styled-0 el-modal-bg"
      id="test-static-id"
    >
      Content within modal
    </div>
  </div>
</DocumentFragment>
`;
